### ISSUE TITLE
Suboptimal plan for single set-returning function in scalar position

### ISSUE THREAD

**Issue by andrioni** [MaterializeInc/materialize#10801]
### What version of Materialize are you using?

v0.20.1-dev (512f92da9)

### How did you install Materialize?

Built from source

### What was the issue?

It seems that we are unable to plan a single set-returning function call in the scalar position as a `FlatMap`, instead using the more complicated implementation with `row_number()` that drastically increases memory utilization/reduces performance.

We assumed this was fine for the Postgres metadata queries we've seen, but now we have users accidentally hitting this issue and seeing dramatically worse performance utilization: here is [one example](https://materializecommunity.slack.com/archives/C015KDVS7EV/p1645326219865269) in Community Slack of a user who hit this issue.

### How to reproduce it

Bootstrapping a simple table with no rows:
```sql
create table srf_json(data jsonb);
```

Sample suboptimal query plan:
```
materialize=> explain typed select jsonb_object_keys(data) from srf_json;
                                                                                      Optimized Plan                                                                                       
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Source materialize.public.srf_json (u1):                                                                                                                                                 +
 | Project (#0)                                                                                                                                                                           +
                                                                                                                                                                                          +
 Query:                                                                                                                                                                                   +
 %0 =                                                                                                                                                                                     +
 | Get materialize.public.srf_json (u1)                                                                                                                                                   +
 | | types = (jsonb?)                                                                                                                                                                     +
 | | keys = ()                                                                                                                                                                            +
 | FlatMap jsonb_object_keys(#0)                                                                                                                                                          +
 | | types = (jsonb?, text)                                                                                                                                                               +
 | | keys = ()                                                                                                                                                                            +
 | Reduce group=(#0)                                                                                                                                                                      +
 | | agg row_number(record_create(list_create(record_create(#0, #1))))                                                                                                                    +
 | | types = (jsonb?, record(?row_number?: bigint,?record?: record(?column?: jsonb?,?column?: text)) list)                                                                                +
 | | keys = ((#0))                                                                                                                                                                        +
 | Project (#1)                                                                                                                                                                           +
 | | types = (record(?row_number?: bigint,?record?: record(?column?: jsonb?,?column?: text)) list)                                                                                        +
 | | keys = ()                                                                                                                                                                            +
 | FlatMap unnest_list(#0)                                                                                                                                                                +
 | | types = (record(?row_number?: bigint,?record?: record(?column?: jsonb?,?column?: text)) list, record(?row_number?: bigint,?record?: record(?column?: jsonb?,?column?: text))?)       +
 | | keys = ()                                                                                                                                                                            +
 | Map if isnull(record_get[0](#1)) then {null} else {record_get[1](record_get[1](#1))}                                                                                                   +
 | | types = (record(?row_number?: bigint,?record?: record(?column?: jsonb?,?column?: text)) list, record(?row_number?: bigint,?record?: record(?column?: jsonb?,?column?: text))?, text?)+
 | | keys = ()                                                                                                                                                                            +
 | Project (#2)                                                                                                                                                                           +
 | | types = (text?)                                                                                                                                                                      +
 | | keys = ()                                                                                                                                                                            +                                                                +
```

Expected query plan:
```
materialize=> explain select key from srf_json, jsonb_object_keys(data) key;
             Optimized Plan
-----------------------------------------
 %0 =                                   +
 | Get materialize.public.srf_json (u98)+
 | FlatMap jsonb_object_keys(#0)        +
 | Project (#1)                         +
```

**Comment by frankmcsherry** [comment]
Pinging @uce to see if this might be appropriate as an Optimization issue.

**Comment by wangandi** [comment]
Edited the suboptimal query plan to include types because it is really not intuitive to me what the expected return type of `row_number(record_create(list_create(record_create(#0, #1)))) ` is.

Looking at the raw plan of the query makes me think that this optimization might be something that should be done in QGM because it is much easier to intuit that row_number() will never be null.
```
explain raw plan for select jsonb_object_keys(data) from srf_json;
                 Raw Plan                 
------------------------------------------
 %0 =                                    +
 | Get materialize.public.srf_json (u1)  +
                                         +
 %1 =                                    +
 | CallTable jsonb_object_keys(#^0)      +
 | Map row_number() over (), #1          +
                                         +
 %2 =                                    +
 | InnerJoin %0 %1 on true               +
 | Map if isnull(#2) then {null} els {#1}+
 | Project (#4)                          +
```

**Comment by andrioni** [comment]
Ah, I might write a bit more about it later on (as it's only described in comments in some places in the codebase), but we have a specific encoding format for both inputs and outputs of window functions, which is what causes this soup of `record_create`/`list_create`/`record_create` in the input to `row_number`, and the `unnest_list` + `record_get` on the output.

**Comment by andrioni** [comment]
Closed by #13120.
